xls/examples/hack.x

```diff
@@ -0,0 +1,256 @@
+// Copyright 2020 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// An incomplete implementation of a the Hack CPU from the https://www.nand2tetris.org/ course.
+//
+// Implemented:
+// - ALU
+// - A/C instructions decoding
+// - RAM addressing
+// - ROM PC addressing
+// Limitation:
+// - missing screen and keyboard memory based IO
+// - smaller RAM and ROM size
+
+fn decode_a_instruction(ins: u16) -> (u15) {
+    let value: u15 = ins[0+:u15];
+    (value,)
+}
+
+#![test]
+fn decode_a_test() {
+    // @21
+    let (value) = decode_a_instruction(u16:0b0000000000010101);
+    let _ = assert_eq(value, u15:21);
+    _
+}
+
+const COMP_0 = u7:0b0101010;
+const COMP_1 = u7:0b0111111;
+const COMP_MINUS_1 = u7:0b0111010;
+const COMP_D = u7:0b0001100;
+const COMP_A = u7:0b0110000;
+const COMP_M = u7:0b1110000;
+const COMP_NOT_D = u7:0b0001101;
+const COMP_NOT_A = u7:0b0110001;
+const COMP_NOT_M = u7:0b1110001;
+const COMP_MINUS_D = u7:0b0001111;
+const COMP_MINUS_A = u7:0b0110011;
+const COMP_MINUS_M = u7:0b1110011;
+const COMP_D_PLUS_1 = u7:0b0011111;
+const COMP_A_PLUS_1 = u7:0b0110111;
+const COMP_M_PLUS_1 = u7:0b1110111;
+const COMP_D_MINUS_1 = u7:0b0001110;
+const COMP_A_MINUS_1 = u7:0b0110010;
+const COMP_M_MINUS_1 = u7:0b1110010;
+const COMP_D_PLUS_A = u7:0b0000010;
+const COMP_D_PLUS_M = u7:0b1000010;
+const COMP_D_MINUS_A = u7:0b0010011;
+const COMP_D_MINUS_M = u7:0b1010011;
+const COMP_A_MINUS_D = u7:0b0000111;
+const COMP_M_MINUS_D = u7:0b1000111;
+const COMP_D_AND_A = u7:0b0000000;
+const COMP_D_AND_M = u7:0b1000000;
+const COMP_D_OR_A = u7:0b0010101;
+const COMP_D_OR_M = u7:0b1010101;
+
+const DEST_NULL = u3:0b000;
+const DEST_M = u3:0b001;
+const DEST_D = u3:0b010;
+const DEST_A = u3:0b100;
+
+const JUMP_NULL = u3:0b000;
+const JUMP_JGT = u3:0b001;
+const JUMP_JEQ = u3:0b010;
+const JUMP_JGE = u3:0b011;
+const JUMP_JLT = u3:0b100;
+const JUMP_JNE = u3:0b101;
+const JUMP_JLE = u3:0b110;
+const JUMP_JMP = u3:0b111;
+
+fn decode_c_instruction(ins: u16) -> (u7, u3, u3) {
+    let comp: u7 = ins[6+:u7];
+    let dest: u3 = ins[3+:u3];
+    let jump: u3 = ins[0+:u3];
+    (comp, dest, jump)
+}
+
+#![test]
+fn decode_c_test() {
+    // MD=D+1
+    let (comp, dest, jump) = decode_c_instruction(u16:0b1110011111011000);
```

Comment:
JFYI you can put underscores to separate nibbles in binary literals if that's easier to read.